src/Microsoft.Extensions.Primitives/StringTokenizer.cs

```diff
@@ -0,0 +1,101 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections;
+using System.Collections.Generic;
+
+namespace Microsoft.Extensions.Primitives
+{
+    /// <summary>
+    /// Tokenizes a <c>string</c> into <see cref="StringSegment"/>s.
+    /// </summary>
+    public struct StringTokenizer :  IEnumerable<StringSegment>
+    {
+        private readonly string _value;
+        private readonly char _separator;
+
+        /// <summary>
+        /// Initializes a new instance of <see cref="StringTokenizer"/>.
+        /// </summary>
+        /// <param name="value">The <c>string</c> to tokenize.</param>
+        /// <param name="separator">The character to tokenize by.</param>
+        public StringTokenizer(string value, char separator)
+        {
+            if (value == null)
+            {
+                throw new ArgumentNullException(nameof(value));
+            }
+
+            _value = value;
+            _separator = separator;
+        }
+
+        /// <inheritdoc />
+        public IEnumerator<StringSegment> GetEnumerator()
+        {
+            return new Enumerator(ref this);
+        }
+
+        IEnumerator IEnumerable.GetEnumerator() => GetEnumerator();
+
+        public struct Enumerator : IEnumerator<StringSegment>
+        {
+            private readonly string _value;
+            private readonly char _separator;
+            private int _index;
+
+            public Enumerator(ref StringTokenizer tokenizer)
+            {
+                _value = tokenizer._value;
+                _separator = tokenizer._separator;
+                Current = default(StringSegment);
+                _index = 0;
+            }
+
+            public StringSegment Current { get; private set; }
+
+            object IEnumerator.Current => Current;
+
+            public void Dispose()
+            {
+            }
+
+            public bool MoveNext()
+            {
+                if (_index > _value.Length)
+                {
+                    return false;
+                }
+
+                var next = _value.IndexOf(_separator, _index);
+                if (next == -1)
+                {
+                    if (_index <= _value.Length)
```

Comment:
Why are this check and the `else` block needed? `_index > _value.Length == false` and `_index <= _value.Length == true` are the same test.